File: Stack_Queue/array_stack.cpp
/* Stack class based on array */
class ArrayStack
{
private:
    vector<int> stack;

public:
    /* Get the length of the stack */
    int size()
    {
        return stack.size();
    }

    /* Determine if the stack is empty */
    bool isEmpty()
    {
        return stack.size() == 0;
    }

    /* Push */
    void push(int num)
    {
        stack.push_back(num);
    }

    /* Pop */
    int pop()
    {
        int num = top();
        stack.pop_back();
        return num;
    }

    /* Access stack top element */
    int top()
    {
        if (isEmpty())
            throw out_of_range("Stack is empty");
        return stack.back();
    }

    /* Return Vector */
    vector<int> toVector()
    {
        return stack;
    }
};

File: java/stack/array_stack.java
import java.util.ArrayList;

/* Stack class based on array */
class ArrayStack {
  private ArrayList<Integer> stack;

  public ArrayStack() {
    // Initialize the list (dynamic array)
    stack = new ArrayList<>();
  }

  /* Get the length of the stack */
  public int size() {
    return stack.size();
  }

  /* Determine if the stack is empty */
  public boolean isEmpty() {
    return size() == 0;
  }

  /* Push */
  public void push(int num) {
    stack.add(num);
  }

  /* Pop */
  public int pop() {
    if (isEmpty()) throw new IndexOutOfBoundsException();
    return stack.remove(size() - 1);
  }

  /* Access stack top element */
  public int peek() {
    if (isEmpty()) throw new IndexOutOfBoundsException();
    return stack.get(size() - 1);
  }

  /* Convert the List to Array and return */
  public Object[] toArray() {
    return stack.toArray();
  }
}
